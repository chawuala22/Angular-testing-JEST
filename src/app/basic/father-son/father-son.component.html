<h5>Hijo</h5>

<code> {{client | json}}</code>

<button *ngIf="client" (click)="onChange(5)" data-test="btnUpdate">
    Cambiar ID
</button>

<button *ngIf="client" (click)="onDelete()" data-test="btnDelete">
    Eliminar cliente
</button>