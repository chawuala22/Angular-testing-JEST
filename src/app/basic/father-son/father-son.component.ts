import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss'],
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onUpdateClient = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }
  onChange(id: number) {
    if (!this.client) return;

    // this.client.id = newId no usar

    this.client = {...this.client,id};

    this.onUpdateClient.emit(this.client);
  }
}
