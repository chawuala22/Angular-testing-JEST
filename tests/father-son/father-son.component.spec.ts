import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FatherSonComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('no deberia aparecer botones si no hay cliente', () => {
    const buttons = compiled.querySelectorAll('Button');
    expect(buttons.length).toBe(0);
  });
  test('no deberia aparecer dos botones si  hay cliente', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('Button');
    expect(buttons.length).toBe(2);
  });

  test('si hay cliente hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('si toco boton borrar cliente emita el evento ondeleteClient', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const buttonDelete = compiled.querySelector('[data-test=btnDelete]');
    buttonDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('Debe emitir onClientUpdate  con el boton de "Cambiar ID"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onUpdateClient, 'emit');

    const buttonChangeId = compiled.querySelector('[data-test=btnUpdate]');
    buttonChangeId?.dispatchEvent(new Event('click'));
    expect(component.onUpdateClient.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan',
    });
  });

  test('Deben emitir onChangeClient con el ID especificado si hay un cliente', () => {
    jest.spyOn(component.onUpdateClient, 'emit');
    component.onChange(10);
    expect(component.onUpdateClient.emit).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onChange(10);
    expect (component.onUpdateClient.emit).toHaveBeenCalledWith({
      id:10,
      name:'Juan'
    })
  });
});
