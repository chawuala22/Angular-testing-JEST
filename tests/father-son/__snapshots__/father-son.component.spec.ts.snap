// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent debe hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
     
  </code>
  
  
</div>
`;

exports[`FatherSonComponent si hay cliente hacer match con el snapshot 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
     
  </code>
  
  
</div>
`;
